db/blob/blob_file_builder.cc

```diff
@@ -372,4 +385,84 @@ void BlobFileBuilder::Abandon(const Status& s) {
   blob_count_ = 0;
   blob_bytes_ = 0;
 }
+
+Status BlobFileBuilder::PutBlobIntoCacheIfNeeded(const Slice& blob,
+                                                 uint64_t blob_file_number,
+                                                 uint64_t blob_offset) const {
+  Status s = Status::OK();
+
+  if (immutable_options_->blob_cache) {
+    bool warm_cache;
+
+    switch (prepopulate_blob_cache_) {
+      case kPrepopulateBlobFlushOnly:
+        warm_cache = (creation_reason_ == BlobFileCreationReason::kFlush);
+        break;
+      case kPrepopulateBlobDisable:
+        warm_cache = false;
+        break;
+      default:
+        // missing case
+        assert(false);
+        warm_cache = false;
+    }
+
+    if (warm_cache) {
+      if (blob_compression_type_ == kNoCompression) {
+        // The cache key does not take into account the real file size. This is
+        // because the blob file in the middle of the flush is unknown to be
+        // exactly how big it is. Therefore, we set the file size here to the
+        // uint64 maximum value to ensure that the warmed cache entries are
+        // available in subsequent inquiries.
+        const OffsetableCacheKey base_cache_key(
+            db_id_, db_session_id_, blob_file_number,
+            std::numeric_limits<uint64_t>::max() /* unknown blob file size */);
```

Comment:
BTW, @ltamasi  We need more research on how to handle this situation, maybe redesign the cache key or pre-populating blobs after flush?

Reply:
So far, yes, I can imagine that when the flush is finished, we might be able to insert the blobs. I'll try it today.

Reply:
It is imperfect to put the blobs into the cache after the flush is complete. This is because we have to use data structure to hold all blobs in memory for each blob file. I'm also not sure if blobs will be freed by the upper call at that time.

Reply:
> Is it possible to precompute the size, or perform a better estimation than std::numeric_limits<uint64_t>::max()?

Options contains an upper limit of blob file size: `mutable_cf_options->blob_file_size`.  Since the exact size is extremely near to it (will be less than or equal to it ), I think we can use it as the file size for the purpose of `OffsetableCacheKey` creation. What do you guys think?

Reply:
Actually, this problem is not unique to blobs: SST file blocks are also preloaded into the block cache as they are generated, i.e. without knowing the final size of the file. The approach taken there is to use a large constant for `max_offset` that still ensures the cache keys have "standard layout" when constructing the `OffsetableCacheKey`. We could take the same approach here. See https://github.com/facebook/rocksdb/blob/main/table/block_based/block_based_table_builder.cc#L899-L904 which in turn is based on https://github.com/facebook/rocksdb/blob/main/cache/cache_key.h#L125-L126

Reply:
It works just as you said. thanks.